### PR TITLE
Add step hooks

### LINKED ISSUE #118: Add before/after step callbacks
**Is your feature request related to a problem? Please describe.**
I have a repetitive set up and teardown code for many step functions.

**Describe the solution you'd like**
- `gobdd.WithBeforeStep(func(ctx gobdd.Context))`
- `gobdd.WithAfterStep(func(ctx gobdd.Context))`

**Describe alternatives you've considered**
Adding a "decorator" to my step functions.

**Additional context**
It would be nice if the default context contains current feature and scenario references.

### PR DESCRIPTION
**Pull request by jirikuncar**
closes #118
partially addresses #121

### CODE REVIEW

File: gobdd.go [hunk 1]
```diff
@@ -116,6 +134,9 @@ type TestingT interface {
 	Run(name string, f func(t *testing.T)) bool
 }
 
+// TestingTKey is used to store reference to current *testing.T instance
+type TestingTKey struct{}
```
Comment (bkielbasa) [inline gobdd.go:138]: this is what I'm not sure about. More handy would be just `ctx.T()`. @sagikazarmark WDYT?

Comment (sagikazarmark) [reply gobdd.go:138]: Why not `consider passing t as argument to scenario hooks`?

Comment (bkielbasa) [reply gobdd.go:138]: yeah, we have 3 options here and I'm not sure which one is a better choice.

Comment (jirikuncar) [reply gobdd.go:138]: > consider passing t as argument to scenario hooks

it's an API change

Comment (sagikazarmark) [reply gobdd.go:138]: I'd go with the one that we already have in place. Having T in both the context and as a parameter is redundant.

Comment (jirikuncar) [reply gobdd.go:138]: > handy would be just `ctx.T()`

It might not be necessary in the future if we start passing `t` to hooks functions in v2.

Comment (jirikuncar) [reply gobdd.go:138]: @sagikazarmark I am trying to find a long term solution here. I would rather change the API for hooks functions in v2 and keep the usage of `TestingT` as not encouraged until we do such change. The API of `gobbd.Context` should be simple and not mix concerns, hence I am not in favor of `ctx.T()`.

Comment (jirikuncar) [reply gobdd.go:138]: > Having T in both the context and as a parameter is redundant.

Hook functions are not accepting `T` as parameter atm.

Comment (sagikazarmark) [reply gobdd.go:138]: I'm not 100% sure I understand which you are advocating.

I think testing.T should be added to hook functions as well. I believe we can do that in a backward compatible manner (see the relocation of context package).

Comment (jirikuncar) [reply gobdd.go:138]: Honestly, I don't know how to change the type signature in backward compatible manner. If you can do it, then I would prefer passing the `t` as argument.

Comment (jirikuncar) [reply gobdd.go:138]: @sagikazarmark @bkielbasa can we merge the current solution for v1 and change the hook function signature in v2?

File: gobdd.go [hunk 2]
```diff
@@ -419,12 +440,28 @@ func (s *Suite) callAfterScenarios(ctx Context) {
...
-	s.callBeforeScenarios(ctx)
-
-	defer s.callAfterScenarios(ctx)
 	t.Run(fmt.Sprintf("%s %s", strings.TrimSpace(scenario.Keyword), scenario.Name), func(t *testing.T) {
+		// NOTE consider passing t as argument to scenario hooks
+		ctx.Set(TestingTKey{}, t)
+		defer ctx.Set(TestingTKey{}, nil)
+
+		s.callBeforeScenarios(ctx)
+		defer s.callAfterScenarios(ctx)
```
Comment (jirikuncar) [inline gobdd.go:463]: ❗ This is an important change. Scenario hooks need access to the current scenario testing object and not the feature one.

File: gobdd.go (hunk 1 above)
Comment (bkielbasa) [reply gobdd.go:138]:
```suggestion
type TestingT struct{}
```

I think the word `Key` here is redundant.

**Comment by bkielbasa** [comment]
One more thing. I think we should document it somehow in the docs. You can find the docs here https://github.com/go-bdd/gobdd/tree/master/docs. For now, you can add the information about the pre-defined key to the context -> https://go-bdd.github.io/gobdd/context.html

**Comment by jirikuncar** [comment]
@bkielbasa I have added documentation for step books and `TestingTKey{}` usage.

File: gobdd.go (hunk 1 above)
Comment (jirikuncar) [reply gobdd.go:138]: @bkielbasa `TestingT` is already defined as interface.

**Comment by sagikazarmark** [comment]
Personally, I would rather see the new hooks having a signature with `testing.T` in them, even if that makes them inconsistent with scenario hooks. We can always find alternative solutions for older code (eg. remove types and check functions runtime).

**Comment by jirikuncar** [comment]
@sagikazarmark the point here is that I would have to change both scenario and step hook signatures here. I am happy start new PR where we can try to find a solution for changing hook signatures.

**Pull request merged**